algebra/direct_limit.lean

```diff
@@ -0,0 +1,275 @@
+import algebra.ring order.lattice
+
+universes u v
+
+theorem congr_arg₂ {α β γ : Type*} (f : α → β → γ) {x₁ x₂ : α} {y₁ y₂ : β}
+  (Hx : x₁ = x₂) (Hy : y₁ = y₂) : f x₁ y₁ = f x₂ y₂ :=
+eq.drec (eq.drec rfl Hy) Hx
+
+open lattice
+
+def is_add_group_hom {α : Type*} {β : Type*} [add_group α] [add_group β] (f : α → β) : Prop :=
+@is_group_hom (multiplicative α) (multiplicative β) _ _ f
+
+attribute [class] is_add_group_hom
+
+namespace is_add_group_hom
+
+variables {α : Type*} {β : Type*} [add_group α] [add_group β] (f : α → β) [hf : is_add_group_hom f]
+
+theorem add (x y) : f (x + y) = f x + f y :=
+@is_group_hom.mul (multiplicative α) (multiplicative β) _ _ f hf x y
+
+theorem zero : f 0 = 0 :=
+@is_group_hom.one (multiplicative α) (multiplicative β) _ _ f hf
+
+theorem neg (x) : f (-x) = -f x :=
+@is_group_hom.inv (multiplicative α) (multiplicative β) _ _ f hf x
+
+end is_add_group_hom
+
+instance is_ring_hom.to_is_add_group_hom {α : Type*} {β : Type*} [ring α] [ring β]
+  (f : α → β) [is_ring_hom f] : is_add_group_hom f :=
+⟨λ _ _, is_ring_hom.map_add f⟩
+
+variables {ι : out_param (Type u)} [inhabited ι] [semilattice_sup ι] {G : ι → Type v}
```

Comment:
Normally a direct limit is defined as being indexed on a directed set, the difference being that we only assume that any two elements have _some_ upper bound, rather than a least upper bound. This is the natural level of generality since it is a sufficient condition for the colimit to be constructed as a quotient of a disjoint union (in a finitary algebraic category like abelian groups).

Would it be too awkward to extend the construction to this setting? I noticed you do use `sup_le` in several places, though I didn't attempt to work out what it is being used for.